### ISSUE TITLE
Building & using zimg on platforms without AVX/FMA

### ISSUE THREAD

**Issue by GrandAdmiralThrawn** [sekrit-twc/zimg#125]
__Problem description:__

When compiling zimg on Linux or Windows+CygWin using GCC 9.2.0 (or any other recent version I presume), the build fails if the instruction set architecture of the host processor does not include ISE's like AVX, AVX2 and FMA.

__How to reproduce:__

Try to compile zimg with GCC on a pre-Sandy Bridge processor, e.g. the first Core i generation (i7-920 to i7-990X or Xeon X55**, X56**) as well as Core 2 chips or older. On the AMD side of things, this should affect all Phenom II and older chips, not perfectly sure about newer ones. Certain modern Intel processors, such as some Pentium Gold chips are also affected. Here is an example of a Coffee Lake generation chip with no AVX/FMA: http://www.cpu-world.com/CPUs/Pentium_Dual-Core/Intel-Pentium%20G5400.html

__How to work around it:__

The person compiling the program needs to do a lot of modifications to the source code to remove all traces of ISE code ≥AVX, which is quite a bit of work.

__Expectation:__

ISE detection at compile time, maybe including some autoconf options to disable the corresponding code paths, excluding them from compilation. So, something like `--disable-avx`, `--disable-avx2`, `--disable-fma`, etc.

Alternatively, parsing of CFLAGS/CXXFLAGS may also be an option, corresponding to the `--march` option or to parameters like `--mno-avx`, `--mno-fma`, etc.

A full transition of SIMD code to nasm assembly with runtime CPU detection and codepath selection may also work, as nasm can assemble AVX/FMA code on older CPUs, but... this is likely unreasonable to ask for.

**Comment by sekrit-twc** [comment]
There is no requirement for the CPU running the compiler to have AVX. Please explain the error you are seeing.

**Comment by GrandAdmiralThrawn** [comment]
Thanks for your reply.

I might have come to a completely wrong conclusion here... Here is an example from the Linux machine using GCC-9.2.0, it's running on a Core i7-980X "Gulftown" CPU, similar to Westmere Xeons:

```
  CXX    src/zimg/colorspace/x86/libavx2_la-operation_impl_avx2.lo
/tmp/ccpr64os.s: Assembler messages:
/tmp/ccpr64os.s:44: Error: suffix or operands invalid for `vpmovzxwd'
/tmp/ccpr64os.s:45: Error: no such instruction: `vgatherdps %ymm4,(%rdi,%ymm0,4),%ymm1'
/tmp/ccpr64os.s:205: Error: no such instruction: `vpbroadcastd %xmm5,%ymm5'
/tmp/ccpr64os.s:212: Error: suffix or operands invalid for `vbroadcastss'
/tmp/ccpr64os.s:213: Error: suffix or operands invalid for `vbroadcastss'
/tmp/ccpr64os.s:256: Error: suffix or operands invalid for `vpmaxsd'
/tmp/ccpr64os.s:257: Error: suffix or operands invalid for `vpminsd'
/tmp/ccpr64os.s:258: Error: no such instruction: `vgatherdps %ymm7,(%rdi,%ymm0,4),%ymm2'
/tmp/ccpr64os.s:505: Error: no such instruction: `vpbroadcastw %xmm0,%xmm0'
make[1]: *** [src/zimg/colorspace/x86/libavx2_la-operation_impl_avx2.lo] Error 1
make[1]: Leaving directory `/home/<censored>/src/zimg-2020-05-8'
make: *** [all-recursive] Error 1
```

I always forget that GNU "as" is not a part of the compiler, but rather of the binutils package. Could it be that it's just my assembler being too old? Really looks like it when I look at that output in more detail. Whoops.

In case I'm correct: Do you know what's the required minimum version of GNU "as" to build zimg?

**Comment by sekrit-twc** [comment]
A Google search suggests that it has worked since at least binutils 2.23 (https://redmine.gromacs.org/issues/1493)

**Comment by GrandAdmiralThrawn** [comment]
Hello again,

It seems your build scripts use `/bin/as` instead of my custom-built `/usr/local/bin/as` (which comes first in my search path, so it's usually the default). I cleaned everything up and did just `export AS=/usr/local/bin/as` before `./configure`, and it worked like a charm. Managed to build it on Linux with custom-built binutils 2.34 and on ancient Windows + Cygwin + custom-built binutils 2.33.1. Verified it by linking ffmpeg 4.2.2 against it and by running a few tests (spatial scaling & tonemapping using the zscale filter). Appears to work just fine!

Sorry for the trouble! This was not a bug after all, but a misconfiguration on my end.

Thank you very much!
